File: modules/core-api/src/main/java/com/enonic/xp/content/site/SiteDescriptor.java
package com.enonic.xp.content.site;


import com.google.common.annotations.Beta;

import com.enonic.xp.form.Form;
import com.enonic.xp.schema.mixin.MixinNames;

@Beta
public final class SiteDescriptor
{
    private final Form form;

    private final MixinNames metaSteps;

    private SiteDescriptor( final Builder builder )
    {
        this.form = builder.form;
        this.metaSteps = builder.metaSteps;
    }

    public Form getForm()
    {
        return form;
    }

    public MixinNames getMetaSteps()
    {
        return metaSteps;
    }

    public static Builder create()
    {
        return new Builder();
    }

    public static Builder copyOf( final SiteDescriptor pageDescriptor )
    {
        return new Builder( pageDescriptor );
    }

    public static class Builder
    {
        private Form form;

        private MixinNames metaSteps;

        private Builder( final SiteDescriptor siteDescriptor )
        {
            this.form = siteDescriptor.getForm();
            this.metaSteps = siteDescriptor.getMetaSteps();
        }

        private Builder()
        {
        }

        public Builder form( final Form form )
        {
            this.form = form;
            return this;
        }

        public Builder metaSteps( final MixinNames metaSteps )
        {
            this.metaSteps = metaSteps;
            return this;
        }

        public SiteDescriptor build()
        {
            return new SiteDescriptor( this );
        }
    }
}

File: modules/admin/admin-ui/src/main/resources/web/admin/common/js/site/SiteDescriptor.ts
module api.site {

    import Form = api.form.Form;
    import MixinNames = api.schema.mixin.MixinNames;

    export class SiteDescriptor implements api.Equitable {

        private form: Form;
        private metaSteps: api.schema.mixin.MixinNames;

        constructor(form: Form, mixinNames: MixinNames) {
            this.form = form;
            this.metaSteps = mixinNames;
        }

        public getForm(): Form {
            return this.form;
        }

        public getMetaSteps(): api.schema.mixin.MixinNames {
            return this.metaSteps;
        }

        static fromJson(json: api.site.json.SiteDescriptorJson): SiteDescriptor {

            return new SiteDescriptor(Form.fromJson(json.form),
                api.schema.mixin.MixinNames.create().fromStrings(json.metaSteps).build());
        }

        public equals(o: api.Equitable): boolean {

            if (!api.ObjectHelper.iFrameSafeInstanceOf(o, SiteDescriptor)) {
                return false;
            }

            var other = <SiteDescriptor>o;

            if (!api.ObjectHelper.equals(this.form, other.form)) {
                return false;
            }

            if (!api.ObjectHelper.equals(this.metaSteps, other.metaSteps)) {
                return false;
            }

            return true;
        }
    }

}